client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -160,8 +162,9 @@ private void releaseSegmentsIfNeeded() throws ReinitializationRequiredException
             log.info("{} releasing segment {}", this, segment);
             SegmentInputStream reader = readers.stream().filter(r -> r.getSegmentId().equals(segment)).findAny().orElse(null);
             if (reader != null) {
-                groupState.releaseSegment(segment, reader.getOffset(), getLag());
-                readers.remove(reader);
+                if (groupState.releaseSegment(segment, reader.getOffset(), getLag())) {
```

Comment:
Say we aren't successful in releasing the segment because the CAS operation failed. Shouldn't we retry or fail with an exception? I'm concerned that we might be silently failing the operation.